### ISSUE TITLE
Wrong action activation and functional loss

### ISSUE THREAD

**Issue by stevenwudi** [Kaixhin/FCN-semantic-segmentation#4]
Hey, thanks for the starter code. I found out the reason why: "This code does not achieve great results (achieves ~40 IoU fairly quickly, but converges there)"

### Following the repo in https://github.com/fyu/drn
In the `main.py`  line 47 I would change the loss to `self.crit = nn.NLLLoss2d(ignore_index=255).cuda()` and  line  70, I would change to `output = F.log_softmax(self.net(input))`.  
Correspondingly the line 67 should change to ` for i, (input, _, target,) in enumerate(train_loader):` that is using the third item of the tuple which is not the one-hot-encoded one.

At least by changing the code, the network continues improving the IOU further (and the change of loss function and output activation make more sense to me).

Hope I made it clear. Cheers

**Comment by Kaixhin** [comment]
Thanks for the suggestions. In the [original FCN paper](https://arxiv.org/abs/1605.06211) the authors use the (independent) sigmoid cross-entropy loss rather than the softmax, as the "softmax operation induces competition between classes and promotes the most confident prediction, but it is not clear that this is necessary or helpful." In any case they also claim that they tried both and it did not seem to make a difference to their results.

I've implemented your changes in the [`nll` branch](https://github.com/Kaixhin/FCN-semantic-segmentation/tree/nll) so please have a look. I've also got results from the current hyperparameters below. This ended after 100 epochs so I've just set of a run for 500 epochs to see if and how much the results improve.
![ious](https://user-images.githubusercontent.com/991891/31760809-d2489b26-b4ad-11e7-96a8-fc0b5190f117.png)

**Comment by Kaixhin** [comment]
At 200 epochs and no noticeable improvement. If I've missed out something then please let me know (or better yet, send a PR).
![ious](https://user-images.githubusercontent.com/991891/31797111-863fb062-b524-11e7-9986-d5c573d3acbb.png)

**Comment by YiwenShaoStephen** [comment]
Hi @Kaixhin , Have you succeed to get improvement on the previous results? And what hyper-parameter did you use? Is it the default value in main.py args? I'm recently also trying to reproduce some decent results with FCN on cityscape, but only achieved under 20% mIOU.

**Comment by Kaixhin** [comment]
@YiwenShaoStephen I haven't looked at this further, but yes the default values are what I ran. I don't PyTorch 0.4 would have broken anything, but I suppose if possible you should try with 0.3 just in case. If you need to adjust hyperparams like batch size you could also possibly see a drop in performance.
